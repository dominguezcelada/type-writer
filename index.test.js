const Typewriter = require('./');

const simpleTypes = new Typewriter()
simpleTypes.addDocument({ "first name": 'Marie', "last name": 'Curie' })
simpleTypes.addDocument({ "first name": 'Ada' })

const complexTypes = new Typewriter()
// union types
complexTypes.addDocument({ foo: null })
complexTypes.addDocument({ foo: 'string' })
complexTypes.addDocument({ foo: 1 })
complexTypes.addDocument({ foo: true })
// empty array
complexTypes.addDocument({ foo: [] })
// single array type
complexTypes.addDocument({ bax: ['string'] }) // array with a single value
// union of array types
complexTypes.addDocument({ foo: [1, 2, 3] })
complexTypes.addDocument({ foo: ['a', 'b', 'c'] })
complexTypes.addDocument({ foo: () => {} })
// nested typing, with a common field
complexTypes.addDocument({ foo: { mandatory: true, bar: 'baz' } })
complexTypes.addDocument({ foo: { mandatory: true, hello: 'world' } })
// empty array with no typing information
complexTypes.addDocument({ bar: [] })
// mixed array
complexTypes.addDocument({ mixed: [1, 2, 'a', 'b'] })

// empty object
complexTypes.addDocument({ empty: {} })

test('Simple test for TypeScript', () => {
  expect(simpleTypes.createInlinedTypeScript()).toMatchSnapshot()
  expect(simpleTypes.createTypeScript({ prefix: 'ProjectName', rootTypeName: 'ProjectName' })).toMatchSnapshot()
});

test('Complex test for TypeScript', () => {
  expect(complexTypes.createInlinedTypeScript()).toMatchSnapshot()
  expect(complexTypes.createTypeScript({ prefix: 'ProjectName', rootTypeName: 'ProjectName' })).toMatchSnapshot()
});

test('Union types for root definition for TypeScript', () => {
  const unionRoot = new Typewriter()
  unionRoot.addDocument({ foo: 'bar' })
  unionRoot.addDocument(['c', 'b', 'c'])
  expect(unionRoot.createInlinedTypeScript()).toMatchSnapshot()
  expect(unionRoot.createTypeScript({ prefix: 'ProjectName' })).toMatchSnapshot()
});

test('Simple test for Vue.js validation', () => {
  expect(simpleTypes.createVueValidation()).toMatchSnapshot()
});

test('Complex test for Vue.js validation', () => {
  expect(complexTypes.createVueValidation()).toMatchSnapshot()
});

test('Simple test for PropTypes', () => {
  expect(simpleTypes.createPropTypes('MyComponent.propTypes ')).toMatchSnapshot()
});

test('Complex test for PropTypes', () => {
  expect(complexTypes.createPropTypes('MyComponent.propTypes ')).toMatchSnapshot()
  console.log(JSON.stringify(complexTypes.types, null, 2))
});

test('Empty object for PropTypes', () => {
  const emptyObject = new Typewriter()
  emptyObject.addDocument({})
  expect(emptyObject.createPropTypes('MyComponent.propTypes ')).toMatchSnapshot()
});
