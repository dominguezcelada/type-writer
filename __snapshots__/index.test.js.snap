// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complex test for PropTypes 1`] = `
"{
  foo: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool,
    PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    ),
    PropTypes.func,
    PropTypes.shape({
      mandatory: PropTypes.bool.isRequired,
      bar: PropTypes.string,
      hello: PropTypes.string
    })
  ]),
  bax: PropTypes.arrayOf(PropTypes.string),
  bar: PropTypes.array,
  mixed: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string])
  ),
  empty: PropTypes.object
}"
`;

exports[`Complex test for TypeScript 1`] = `
"{
  foo?:
    | null
    | string
    | number
    | boolean
    | Array<number | string>
    | { (): any }
    | {
        mandatory: boolean,
        bar?: string,
        hello?: string
      },
  bax?: Array<string>,
  bar?: Array<any>,
  mixed?: Array<number | string>,
  empty?: {}
}"
`;

exports[`Complex test for TypeScript 2`] = `
"type ProjectNameFoo = {
  mandatory: boolean,
  bar?: string,
  hello?: string
}

type ProjectNameEmpty = {}

type ProjectName = {
  foo?:
    | null
    | string
    | number
    | boolean
    | Array<number | string>
    | { (): any }
    | ProjectNameFoo,
  bax?: Array<string>,
  bar?: Array<any>,
  mixed?: Array<number | string>,
  empty?: ProjectNameEmpty
}
"
`;

exports[`Complex test for Vue.js validation 1`] = `
"{
  foo: [String, Number, Boolean, Array, Function, Object],
  bax: Array,
  bar: Array,
  mixed: Array,
  empty: Object
}"
`;

exports[`Empty object for PropTypes 1`] = `""`;

exports[`Named types 1`] = `
"type ProjectNamePayloadIssue = {
  title: string
}

type ProjectNamePayload = {
  issue: ProjectNamePayloadIssue
}

type ProjectName = {
  payload: ProjectNamePayload
}
"
`;

exports[`Simple test for PropTypes 1`] = `
"{
  \\"first name\\": PropTypes.string.isRequired,
  \\"last name\\": PropTypes.string
}"
`;

exports[`Simple test for TypeScript 1`] = `
"{
  \\"first name\\": string,
  \\"last name\\"?: string
}"
`;

exports[`Simple test for TypeScript 2`] = `
"type ProjectName = {
  \\"first name\\": string,
  \\"last name\\"?: string
}
"
`;

exports[`Simple test for Vue.js validation 1`] = `
"{
  \\"first name\\": { type: String, required: true },
  \\"last name\\": String
}"
`;

exports[`Union types for root definition for TypeScript 1`] = `
"| {
      foo: string
    }
  | Array<string>"
`;

exports[`Union types for root definition for TypeScript 2`] = `
"type ProjectName = {
  foo: string
}

type Root = ProjectName | Array<string>
"
`;
